// photofinish/dist/lib/resultsExporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.exportResults = void 0;
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const fs_1 = tslib_1.__importDefault(require("fs"));
const nodeVersionUtils_1 = require("./internal/nodeVersionUtils");
const filenameUtils_1 = require("./internal/filenameUtils");
function exportResults(benchmarkResults, options = {}) {
    var _a;
    const exportPath = (_a = options === null || options === void 0 ? void 0 : options.exportPath) !== null && _a !== void 0 ? _a : './results';
    const exportData = {
        runtimeVersion: nodeVersionUtils_1.runtimeVersion,
        benchmarkName: benchmarkResults.benchmarkName,
        benchmarkEntryName: benchmarkResults.benchmarkEntryName,
        benchmarkEntryVersion: benchmarkResults.benchmarkEntryVersion,
        benchmarkCycles: benchmarkResults.benchmarkCycles,
        benchmarkCycleSamples: benchmarkResults.benchmarkCycleSamples,
        label: benchmarkResults.label,
        warmupCycles: benchmarkResults.warmupCycles,
        meanTimeNs: benchmarkResults.meanTime.getTimeInNanoSeconds(),
        meanTimeMs: benchmarkResults.meanTime.getTimeInMilliSeconds(),
    };
    const filename = filenameUtils_1.generateResultsFilename(benchmarkResults);
    const resolvedPathFile = path_1.default.resolve(exportPath, filename);
    fs_1.default.mkdirSync(exportPath, { recursive: true });
    fs_1.default.writeFileSync(resolvedPathFile, JSON.stringify(exportData, null, 2));
}
exports.exportResults = exportResults;
