// photofinish/dist/lib/resultsImporter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadResults = void 0;
const { loadFiles } = require('load-goblin');
async function loadResults(baseResultsDir) {
    const loadResults = await loadFiles({
        filter: '*.json',
        rootDirectory: baseResultsDir,
        recursively: true,
    });
    return loadResults.map((entry) => {
        return {
            runtimeVersion: entry.content.runtimeVersion,
            benchmarkName: entry.content.benchmarkName,
            benchmarkEntryName: entry.content.benchmarkEntryName,
            benchmarkEntryVersion: entry.content.benchmarkEntryVersion,
            meanTimeNs: entry.content.meanTimeNs,
            meanTimeMs: entry.content.meanTimeMs,
            warmupCycles: entry.content.warmupCycles,
            benchmarkCycles: entry.content.benchmarkCycles,
            benchmarkCycleSamples: entry.content.benchmarkCycleSamples,
            label: entry.content.label,
        };
    });
}
exports.loadResults = loadResults;
